package/src/core/with-plugins.ts

```diff
@@ -21,7 +21,13 @@ export const withPlugins = <TPlugins extends NonEmptyArray<AnyPlugin>>({
 	hooks: providedHooks,
 }: { name: string; plugins: TPlugins; hooks: ExtendedHooks<TPlugins> }) => {
 	// Overrides plugins with same name
-	const resolvedPlugins = Object.values(Object.fromEntries(plugins.map(plugin => [plugin.name, plugin])))
+	const resolvedPlugins = Object.values(
+		Object.fromEntries(
+			plugins.map((plugin) => [plugin.name, plugin.setup({ name })]),
+		),
+	) as Array<
+		Partial<Record<keyof Hooks, (params: any) => Record<string, unknown>>>
+	>;
```

Comment:
I don't think `Object.fromEntries` followed by `Object.values` is guaranteed to maintain the same order as the input. Without this guarantee, we can't provide plugin overrides correctly.

The `Map` class is specified to provide this guarantee. The current implementation in v8 will guarantee only up to 6 entries for objects. After that, it will depend on other objects previously created to reuse some internals for large objects.

---

It seems like the goal here is to deduplicate plugins with the same name. I think it would be better to do that with just an array to avoid losing the ordering or depending on implementation details:

```suggestion
	// Overrides plugins with same name, keeping only the last occurrence
	const resolvedPlugins = plugins
		.filter((plugin, index, self) => self.findLastIndex(other => other.name === plugin.name) === index)
		// Setup plugins with the integration parameters
		.map((plugin): Partial<Record<keyof Hooks, (params: any) => Record<string, unknown>>> => plugin.setup({ name }));
```

Reply:
I went into a trip through a rabbit hole because of this.

MDN docs for [`Object.values`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values#description) states that ordering of the values is the same as the keys iterated over by the `for...in`.

MDN docs for `for...in` iteration states [^1]:

> The traversal order, as of modern ECMAScript specification, is well-defined and consistent across implementations. Within each component of the prototype chain, all non-negative integer keys (those that can be array indices) will be traversed first in ascending order by value, then other string keys in ascending chronological order of property creation.

[^1]: 3rd paragraph of https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#description

So, I decided to read the spec to understand if that is a proper guarantee in our case or not. The spec for key enumeration as used by `Object.values` doesn't have a very promising start [^2]:

> The mechanics and order of enumerating the properties is not specified but must conform to the rules specified below.

[^2]: https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-enumerate-object-properties

The following rules are not _directly_ related to property order, but the very last detail in the rules list is:

> In addition, if neither O nor any object in its prototype chain is a [Proxy exotic object](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#proxy-exotic-object), [TypedArray](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#typedarray), [module namespace exotic object](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#module-namespace-exotic-object), or implementation provided [exotic object](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#exotic-object), then the iterator must behave as would the iterator given by [CreateForInIterator](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-createforiniterator)(O) until one of the following occurs:
> [... some corner cases ...]

Although it follows an explicit statement that the ordering is not specified, the `CreateForInIterator` uses an algorithm [^3] that will yield the same order as [`OrdinaryOwnPropertyKeys`](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-ordinaryownpropertykeys), which is the ordering mentioned by MDN. So, it can be interpreted as the ordering being well-defined for all regular JS objects (and all other objects that don't fall in any of those many exception cases).

But then, right after that very same rule, there is:

> ECMAScript implementations are not required to implement the algorithm in [14.7.5.10.2.1](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-%foriniteratorprototype%.next) directly. They may choose any implementation whose behaviour will not deviate from that algorithm unless one of the constraints in the previous paragraph is violated.

It seems to be up to interpretation whether an algorithm that goes through the exact same values as 14.7.5.10.2.1 but in a different order would be a deviation. The provided algorithm yields the keys in a stable order, but at the same time, the order is specified as not being part of the specification.

One hour I spent trying to understand the super convoluted algorithm implemented by v8, which is split into several files. It is _not_ the algorithm provided in the spec, and I found one divergence. The v8 implementation breaks chronological order if a property is removed and re-added to the same object (or at least it interprets what that ordering means differently).

[^3]: 14.7.5.10.2.1: https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-%foriniteratorprototype%.next

---

In conclusion, the current code is safe here since this specific case will remain in the same order in v8. According to the MDN docs, this is consistent with other runtimes (although that is not really relevant since Astro only supports Node).

But I still think the `.filter().map()` is cleaner and conveys the intent more clearly. It is also faster for a small number of plugins, which will probably be the norm.

Reply:
I'm impressed by your research, thanks a lot! This comment alone is worth sharing in `#showcase`!